src/utils.c

```diff
@@ -2497,3 +2497,151 @@ gchar *utils_get_os_info_string(void)
+ *
+ * @note Invalid characters in the needles string will be ignored
+ *
+ * @note This function uses the global variable app (type GeanyApp)
+ *
+ * @note Substituting "%%e" for a file called fam.c would give the result "fam.cam".
+ *       To prevent this side-effect, multiple consecutive "%" are reduced to "%".
+ */
+gchar *utils_replace_placeholder(const GeanyDocument *doc, const gchar *src, const gchar *needles)
+{
+	GString *haystack    = NULL;
+	gchar   *executable  = NULL;
+	gchar   *replacement = NULL;
+	gint    line_num     = 0;
+	GRegex  *regex       = NULL;
+	GString *errormsg    = NULL;
+
+	g_return_val_if_fail(doc != NULL    , NULL);
+	g_return_val_if_fail(doc->is_valid  , NULL);
+	g_return_val_if_fail(needles != NULL, NULL);
+	g_return_val_if_fail(src != NULL    , NULL);
+
+	haystack = g_string_new(src);
+
+	/* Reduce "%%%..." to one "%" */
+	regex = g_regex_new("%+", 0, 0, NULL);
+
+	replacement = g_regex_replace(regex, haystack->str, -1, 0, "%", 0, NULL);
+	g_string_assign(haystack, replacement);
+
+	g_regex_unref(regex);
+	regex = NULL;
+	g_free(replacement);
+	replacement = NULL;
+
+	/* Execute effective substitutions */
+
+	if (app->project) {
+		/* replace %p with the current project's (absolute) base directory */
+		replacement = project_get_base_path();
+		utils_string_replace_all(haystack, "%p", replacement);
+	}
+
+	if (doc->file_name != NULL)
+	{
+		if (g_utf8_strchr(needles, -1, 'd') != NULL) {
+			/* replace %d with the absolute path */
+			replacement = g_path_get_dirname(doc->file_name);
+			utils_string_replace_all(haystack, "%d", replacement);
+			g_free(replacement);
+		}
+
+		if (g_utf8_strchr(needles, -1, 'e') != NULL) {
+			/* replace %e with the filename (excluding extension) */
+			executable = utils_remove_ext_from_filename(doc->file_name);
+			replacement = g_path_get_basename(executable);
+			utils_string_replace_all(haystack, "%e", replacement);
+			g_free(replacement);
+			g_free(executable);
+		}
+
+		if (g_utf8_strchr(needles, -1, 'f') != NULL) {
+			/* replace %f with the filename (including extension) */
+			replacement = g_path_get_basename(doc->file_name);
+			utils_string_replace_all(haystack, "%f", replacement);
+			g_free(replacement);
+		}
+
+		if (g_utf8_strchr(needles, -1, 'l') != NULL) {
+			/* replace %l with the current 1-based line number */
+			line_num = sci_get_current_line(doc->editor->sci) + 1;
+			replacement = g_strdup_printf("%i", line_num);
+			utils_string_replace_all(haystack, "%l", replacement);
+			g_free(replacement);
+		}
+
+		/* if app->Project then                      */
+		/*   all %p placeholders are already removed */
+		/* else                                      */
+		/*   fall back to absolute path/name.ext     */
+		if (g_utf8_strchr(needles, -1, 'p') != NULL) {
+			/* replace %p with the absolute path/filename (including extension) */
+			utils_string_replace_all(haystack, "%p", doc->file_name);
```

Comment:
This doesn't make any sense, and it used to be equivalent to `%d`, not `%d/%f`.
As is `%p` is entirely unreliable as it's impossible to know whether it's gonna be a path to a directory or a file. And it breaks use of `%p` in build commands.

Reply:
#4318